### PR TITLE
Allowed post-processors to be used in functor parameters

### LINKED ISSUE #25012: Allow post-processors to be used in functor parameters
## Reason
It's sometimes useful to pass post-processors in for parameters of type `MooseFunctorName`.

## Design
This would allow post-processors to be used in parameters of type `MooseFunctorName`.

## Impact
No change to existing input files. Greater flexibility.

### PR DESCRIPTION
**Pull request by joshuahansel**
Closes #25012

### CODE REVIEW

File: framework/src/postprocessors/Postprocessor.C [hunk 1]
```diff
@@ -64,3 +65,101 @@ Postprocessor::getValue() const
...
+Postprocessor::evaluateGradient(const ElemPointArg & /*elem_point_arg*/,
+                                const Moose::StateArg & /*state*/) const
+{
+  return 0;
+}
+
+typename Postprocessor::DotType
+Postprocessor::evaluateDot(const ElemArg & /*elem_arg*/, const Moose::StateArg & /*state*/) const
+{
+  return 0;
```
Comment (GiudGiud) [inline framework/src/postprocessors/Postprocessor.C:138]: I d rather an error for time derivatives
or a doOnce warning at least

Comment (joshuahansel) [reply framework/src/postprocessors/Postprocessor.C:138]: Yeah, a doOnce sounds like a good idea. I think I'll do the same for the gradient.

Comment (GiudGiud) [reply framework/src/postprocessors/Postprocessor.C:138]: no not gradient imo
a PP is assumed to have a 0 gradient, that's ok

**Comment by joshuahansel** [comment]
Hm, I was assuming that this would give me exactly one PP in `FEProblemBase::addPostprocessor`:
```
    theWarehouse()
        .query()
        .template condition<AttribName>(name)
        .template condition<AttribSystem>("UserObject")
        .template condition<AttribThread>(tid)
        .queryInto(uos);
```
but apparently not when run with multiple threads. I thought `.condition<AttribThread>(tid)` would take care of this.

**Comment by joshuahansel** [comment]
Nevermind, it was because of _zero_ matches, not multiple.

**Comment by joshuahansel** [comment]
I'm struggling with `FEProblemBase::addPostprocessor` - I'm doing something wrong with respect to threads. For example, when running my new test with 2 threads:
```
functors/get_functor.test: *** ERROR ***
functors/get_functor.test: The following error occurred in the object "MOOSE Problem", of type "FEProblem".
functors/get_functor.test: 
functors/get_functor.test: No functor ever provided with name 'testpp', which was requested by 'get_pp'.
```
Do I need the line like `this->addFunctor(name, *functor, tid);` for `tid` such that nothing is returned by my query?

**Comment by loganharbour** [comment]
Do we not want to get the most recently calculated value (as the docs say) instead of the result of `getValue()`? The most recently saved value is stored in the reporter system

**Comment by GiudGiud** [comment]
I think getValue() returns the most recent value, stored in _value very often.
I think in the future we may want to have a boolean on the PP functor like "calculate PP functor on-the-fly" that would re-calculate it. but for right now having getValue() is great!

**Comment by joshuahansel** [comment]
I agree, I'd think `getValue()` should always return the most recent value.

**Comment by loganharbour** [comment]
In many cases it does not return the value computed at the last execute flag. I don't think it's appropriate here.

**Comment by GiudGiud** [comment]
That's an issue with those PPs' implementation of getValue() though?
What would be the alternative to getValue for Josh's PR?

**Comment by loganharbour** [comment]
> That's an issue with those PPs' implementation of getValue() though?

Sure, but we have an option to be consistent so why should we not? We should be using the same values that everyone gets from the PP interface

> What would be the alternative to getValue for Josh's PR?

https://github.com/idaholab/moose/blob/cdb4886e5078301450e157460d7f026822cfaf1a/framework/src/postprocessors/Postprocessor.C#L47

Store this as a reference in the postprocessor. Setup a const method to get it. Rather simple

**Comment by GiudGiud** [comment]
Seems fine. Maybe getValue() should disappear then, it s a duplicate, and we force people to finalize their value on finalize()

**Comment by joshuahansel** [comment]
> In many cases it does not return the value computed at the last execute flag.

I still don't understand how this is possible. Let's assume that `getValue()` is already const.

I think what you're proposing is fine, but I agree with Guillaume that it seems to imply that the separation of `getValue()` and `finalize()` is kind of pointless, if we're never allowed to assume that `getValue()` reflects the last executed `finalize()`.

**Comment by loganharbour** [comment]
> I still don't understand how this is possible. Let's assume that getValue() is already const.

Doesn't matter if it's const or not. In a lot of cases, it returns a value computed by something else _at the point at which it is called_.

> I think what you're proposing is fine, but I agree with Guillaume that it seems to imply that the separation of getValue() and finalize() is kind of pointless, if we're never allowed to assume that getValue() reflects the last executed finalize().

The problem is that `finalize()` is a void function because it comes from `UserObject`. It'll never return a value. Sure - long term we can change the behavior to have someone say `_value = ...` in `finalize()`, but that's a big change. This is just an artifact of the API that was chosen. By only using the declared value everywhere (instead of `getValue()`), we can at least guarantee that we're doing the right thing even though the API is poor.

**Comment by loganharbour** [comment]
I know that it's annoying, but my goal here is to at least future proof us to regolding and making your statement about the execute on timing in documentation be _completely_ valid

**Comment by joshuahansel** [comment]
> by something else

Ok, thanks, this part clarifies it for me. Given this possibility, I think your suggestion is good, so I'll do this.

I probably did the whole `getValue() const` thing for nothing (my fault obviously) 😢

**Comment by loganharbour** [comment]
Good with me now on the value implementation. I added at least 20 lines of docstring comments, so I expect at least multiple change requests for those

File: framework/include/postprocessors/Postprocessor.h [hunk 2]
```diff
@@ -10,39 +10,119 @@
...
 
 protected:
-  const std::string _pp_name;
+  /// MOOSE object
+  const MooseObject & _pp_moose_object;
+
+  /// Post-processor name
+  const std::string & _pp_name;
+
+  /// The current value, which is the Reporter value that changes when we execute UOs in the problem)
```
Comment (joshuahansel) [inline framework/include/postprocessors/Postprocessor.h:81]:
```suggestion
  /// The current value, which is the Reporter value that changes when we execute UOs in the problem
```

**Comment by joshuahansel** [comment]
@loganharbour Thanks for the help. I was looking for a good way to get the reporter value as a constant reference - didn't see that the declare method gave one. As for the other changes to `FEProblemBase`, I don't really understand why I had to do what I did before, but it's what I found that I had to do.

File: framework/include/postprocessors/Postprocessor.h (hunk 2 above)
Comment (joshuahansel) [inline framework/include/postprocessors/Postprocessor.h:81]:
```suggestion
  /// The current value, which is the Reporter value that changes when we execute UOs in the problem
```

**Comment by loganharbour** [comment]
We should probably test all of the functor calls for a postprocessor

File: framework/doc/content/syntax/Functors/index.md [hunk 3]
```diff
@@ -17,6 +18,11 @@ but for `Functors`, the space arguments can be an element, a point in an element
 a face of an element, and so on. The time arguments represent the state of the functor : current, previous value (whether in time or iteration),
 or value before that previous one.
 
+!alert warning title=Using post-processors
+Note that when using post-processors, instead of being evaluated on-the-fly at
+a point in space and time, the most recently computed value is used, in accordance
+with the `execute_on` for the post-processor.
+
```
Comment (GiudGiud) [inline framework/doc/content/syntax/Functors/index.md:25]: add a note to the PP documentation saying it inherits the functor interface. Functions and variables doc should have a note already

File: framework/include/postprocessors/Postprocessor.h [hunk 4]
```diff
@@ -10,39 +10,119 @@
...
   /**
    * Returns the name of the Postprocessor.
    */
-  std::string PPName() const { return _pp_name; }
+  const std::string & PPName() const { return _pp_name; }
 
 protected:
-  const std::string _pp_name;
+  /// MOOSE object
+  const MooseObject & _pp_moose_object;
```
Comment (GiudGiud) [inline framework/include/postprocessors/Postprocessor.h:76]: umm this should probably be private? we dont want users working with this do we?

Comment (loganharbour) [reply framework/include/postprocessors/Postprocessor.h]: There's no harm as it's typically the same object you're in, but I do like the idea of making it private to get rid of the clutter

**Comment by joshuahansel** [comment]
Comments should be addressed now.

**Review by loganharbour — COMMENTED** [review]
I approve of the content that isn't my own.

@GiudGiud or @lindsayad please

**Review by GiudGiud — APPROVED** [review]

**Comment by GiudGiud** [comment]
Excellent. This will serve us well on the FV side as well!

**Pull request merged**